### DISCUSSION TITLE
Use rabbitmq direct reply-to for RPC calls

### DISCUSSION THREAD

**Discussion by bodograumann** [Lancetnik/Propan#69] [Ideas]
The use case of [RPC over MQ][1] is very interesting, especially for testing purposes.
Looking at the code, it seems a [dedicated reply queue][2] is created for each request.
I think that we can avoid that by using an existing reply queue or (at least for RabbitMQ) by using the  [Direct Reply-To feature][3].

Implementing it should be simple.
`broker.publish` already accepts a `reply_to` argument and it is currently applied to the message.
Now if both `callback=True` and `reply_to` are set, we could avoid creating a dedicated reply queue and use the one from `reply_to`.
In particular `broker.publish(queue=…, callback=True, reply_to="amq.rabbitmq.reply-to")` would be using the Direct Reply-To feature of RabbitMQ.

At the moment, when using`callback` and `reply_to` at the same time, the latter would override the former, but propan would still be waiting for a reply on the anonymous reply queue it created internally. So the reply can never be received in this situation.

I also noticed that the `callback_queue` is currently not deleted. This should probably be added as well, when it is created internally. (When given by the user via `reply_to` it should be kept of course. `amq.rabbitmq.reply-to` isn't even a real queue and cannot be deleted.)

[1]: https://lancetnik.github.io/Propan/getting_started/4_broker/5_rpc/#client
[2]: https://github.com/Lancetnik/Propan/blob/main/propan/brokers/rabbit/rabbit_broker.py#L180
[3]: https://www.rabbitmq.com/direct-reply-to.html

**Comment by Lancetnik** [comment]
@bodograumann thanks for notice that `callback` overrides original `reply_to`. Obviously, it should be fixed.

> it seems a dedicated reply queue is created for each request.

It is exactly how **RabbitMQ** implements [Direct Reply-To feature](https://www.rabbitmq.com/direct-reply-to.html).
Take a look at python or any other language [example](https://www.rabbitmq.com/tutorials/tutorial-six-python.html).
In all cases anonymous exclusive queue is created.

> I also noticed that the callback_queue is currently not deleted.

Queues, declared as `exclusive` automatically deleting after disconnect. So, if you connect only to send a message, it will be deleted imidiatelly. Another case, it will be deleted at reconnection (every 30 minutes with default **RabbitMQ** settings).
I will test fake-queue behavior tomorrow, but for now I don't see a problem. I am just following the **RabbitMQ RPC** tutorial, linked by you.

**Reply by bodograumann** [reply]
> > I also noticed that the callback_queue is currently not deleted.
> 
> Queues, declared as `exclusive` automatically deleting after disconnect. So, if you connect only to send a message, it will be deleted imidiatelly. Another case, it will be deleted at reconnection (every 30 minutes with default **RabbitMQ** settings). I will test fake-queue behavior tomorrow, but for now I don't see a problem. I am just following the **RabbitMQ RPC** tutorial, linked by you.

Ah, thanks for the explanation. That makes total sense.

> It is exactly how **RabbitMQ** implements [Direct Reply-To feature](https://www.rabbitmq.com/direct-reply-to.html). Take a look at python or any other language [example](https://www.rabbitmq.com/tutorials/tutorial-six-python.html). In all cases anonymous exclusive queue is created.

Please have another look at the documentation I linked: https://www.rabbitmq.com/direct-reply-to.html
Creating a new queue for every call has quite a bit of overhead, so often there is a long-lasting queue for the replies. To avoid having to manage that queue, RabbitMQ provides some magic if you use the special name "amq.rabbitmq.reply-to".
I think supporting both those use cases (explicit long-lasting reply queue and magic virtual queue) by letting the user provide the `reply_to` argument makes a lot of sense.

**Reply by Lancetnik** [reply]
One more time: declaring queue without name is exactly the thing, recommended by **RebbitMQ** to implement "amq.rabbitmq.reply-to" from your acticle. This way you declare not a really queue, but virtual queue to get one response.
In **Propan** case user is able to set `reply_to` by himself to get response in long-libing queue, or just use `callback=True` to get response with virtual one.

**Reply by bodograumann** [reply]
You are right that declaring a short-lived anonymous queue is one way to configure RPC over MQ.
However I am talking about a different method.
Quoting from the article:

> To use direct reply-to, an RPC client should:
>
> Consume from the pseudo-queue amq.rabbitmq.reply-to in no-ack mode. There is no need to declare this "queue" first, although the client can do so if it wants.
> Set the reply-to property in their request message to amq.rabbitmq.reply-to.
> The RPC server will then see a reply-to property with a generated name.

That's why there is the need for setting `reply_to` to `amq.rabbitmq.reply-to` for this method to work.

**Reply by Lancetnik** [reply]
Man, but I did it [here](https://github.com/Lancetnik/Propan/blob/main/propan/brokers/rabbit/rabbit_broker.py#L351)
Anyway you need to have a queue object to consume from. You can't just send a message with `reply_to` and consume from nothing.
> However I am talking about a different method.
The method you talking about is exaclty what I am doing. Also it is a reccomended by **RabbitMQ** way. Let's try to implement something else. I am glad t merge your PR if you can do it,

**Reply by Lancetnik** [reply]
But, you are right: management plugin show that `anonymous` queue as existed. It was deleted only with channel disconnected.
But, for now, I have no ideas how to implement `pseudo-queue` from the original article.
When I was tried to get message by `reply_to` name direclty without declaring **RabbitMQ** raised error `QueueNotExists`

**Reply by bodograumann** [reply]
True, you use the anonymous queue you create for just this one call in the reply-to property.

The RabbitMQ docs say:
> Where does the client's queue come from? The client can declare a single-use queue for each request-response pair. But this is inefficient; even a transient unreplicated queue can be expensive to create and then delete (compared with the cost of sending a message). This is especially true in a cluster as all cluster nodes need to agree that the queue has been created, agree on its type, replication parameters, and other metadata.

There are two other ways to go about though.
The user could already have created a reusable reply queue. One difficulty would be when multiple requests go in parallel over it. I honestly don't know how this is usually handled in detail. Probably with the correlationId.

From the article again:
> So alternatively the client can create a long-lived queue for its replies. But this can be fiddly to manage, especially if the client itself is not long-lived.

So they propose another option:
> The direct reply-to feature allows RPC clients to receive replies directly from their RPC server, without going through a reply queue.

So when using this "Direct Reply-To"-Feature, there is no actual reply queue created. RabbitMQ handles it all in the background. As I understand it:
- The client sets `reply-to` to `amq.rabbitmq.reply-to` and listens to this pseudo-queue,
- the server gets a message with a one-off `reply-to` e.g. `foobar-request-1234`,
- the server publishes the reply to the default exchange with pseudo-queue name `foobar-request-1234` as routing key and
- the message will automatically be relayed to the correct client by rabbitmq. No actual reply queue necessary.

So you are absolutely correct that from the clients point of view it still looks like there is some callback queue he consumes from.
I have to admit that I didn't completely think this through. Instead of declaring a new `callback_queue`, we need to consume from `amq.rabbitmq.reply-to` to support Direct Reply-To.
I am not completely sure what the best interface is here. We could
- look at the `reply_to` argument and use `callback_queue = channel.declare_queue(name=reply_to, passive=True)` if it is set,
- define a new argument where the user can pass a RabbitQueue object as reply queue (and that could then contain the Direct Reply-To pseudo queue if desired) or
- allow passing this reply RabbitQueue object in some existing argument like `reply_to` or `callback`.

One other detail is, that for Direct Reply-To the client needs to consume in auto-ack mode. Note sure if we do that currently for RPC.

**Reply by bodograumann** [reply]
> But, you are right: management plugin show that `anonymous` queue as existed. It was deleted only with channel disconnected. But, for now, I have no ideas how to implement `pseudo-queue` from the original article. When I was tried to get message by `reply_to` name direclty without declaring **RabbitMQ** raised error `QueueNotExists`

Sorry I didn't see that answer. Hope what I wrote above helps figuring out how to do Direct Reply-To without declaring a completely new queue. :-)

**Reply by Lancetnik** [reply]
I read this article too. Looks great, but it seems like I can publish message with some uuid as `reply_to` and **RabbitMQ** creates pseudo-queue binded to default exchange with the same name. But it doesn't.
I tried a few scenarios:
1) publish message with random `reply_to`, get the pseudo-queue by name - error "Queue not found"
2) publish message with random `reply_to`, try to get message from default exchange by queue name - error "Queue not found"
3) declare anonymous queue, publish message with `reply_to` to it  - real queue was declared

Looks like there is no interfaces to implement *Direct Reply-To* or I can't find it.

**Reply by bodograumann** [reply]
Have you tried using the exact string `"amq.rabbitmq.reply-to"` as `reply_to` when publishing?

**Reply by Lancetnik** [reply]
Wait a second... Looks like I got it. There are too few materials on the Internet to deal with it so simply.

**Reply by Lancetnik** [reply]
But I don't undestand how get exactly my response from `"amq.rabbitmq.reply-to"` yet

**Comment by Lancetnik** [comment]
Wait: you can't use `callback=True` and `reply_to` both the same time. Callback should be used then you need to get response right here. `reply_to` should be used then you need to get response in long-living queue with handler, declared on your app.
So, it looks lika candidate to `ValueError`, not a wrong behavior.

**Comment by Lancetnik** [comment]
Also, please *do not use RPC to testing*. It's a production feature.
I was working hard on [local testing](https://lancetnik.github.io/Propan/getting_started/7_testing/), so, please, use it with pytest locally and in CI.

**Reply by bodograumann** [reply]
Yes definitely for unit testing and such I will use the testnig facilities which won't require all the dependencies.
However migrating to Propan requires changes and iteration on how the service interacts with RabbitMQ.
That is why I want to have a minimal RPC over MQ client that checks whether the service replies correctly with the real RabbitMQ server.

**Reply by Lancetnik** [reply]
Sure, thanks for your trust in this young project. We can make it greater together with my job and your production feedback!

**Comment by Lancetnik** [comment]
@bodograumann thanks for your help. RPC over MQ become better by you.

**Reply by bodograumann** [reply]
Thank you so much for implementing this. Just got server and client working with propan.
First I was a bit confused, because I did not see any reply queue in rabbitmq,
but then I saw that we are now using the Direct Reply-To method *by default* — Fantastic! :heart:

**Reply by Lancetnik** [reply]
I am glad to improve this project together. But, this functionality has some limitations you should know: publish with reply_to is synchronous per RabbitBroker instanse by asyncio.Lock. This limitation is connected with **RabbitMQ** amqp.reply_to single consumer limitation. So, if you sending messages without creating asyncio.Task for each, It shouldn't be a problem

**Comment by Lancetnik** [comment]
Well, now we have another problem: miltiple consuming from reply_to is forbidden. So, RPC request is blocked by another request...

**Reply by Lancetnik** [reply]
So, we can use multiple RPC requests from different connections or from one at sync mode. But if we try to create multiple requests at one connection it caused to error.
